algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,199 @@
+package com.twitter.algebird
+
+/**
+ * Exponential Histogram algorithm from
+ * http://ilpubs.stanford.edu:8090/504/1/2001-34.pdf.
+ *
+ *
+ * Next step - code up the l-canonical representation of numbers,
+ * then do some of the fancier business in the paper!
+ */
+object ExpHist {
+  case class Config(k: Int, windowSize: Long) {
+    // Maximum number of buckets of size 2^i allowed in the repr of
+    // this exponential histogram.
+    def maxBuckets: Int = math.ceil(k / 2.0).toInt + 1
+
+    // Returns the last timestamp before the window. any ts <= [the
+    // returned timestamp] is outside the window.
+    def expiration(currTime: Long): Long = currTime - windowSize
+  }
+
+  def empty(k: Int, windowSize: Long): ExpHist =
+    ExpHist(Config(k, windowSize), Vector.empty, 0L)
+}
+
+case class ExpHist(conf: ExpHist.Config, buckets: Vector[BucketSeq], time: Long) {
+  /**
+   * Returns the same ExpHist with a new window. If the new window is
+   * smaller than the current window, evicts older items.
+   */
+  def withWindow(newWindow: Long): ExpHist =
+    copy(conf = conf.copy(windowSize = newWindow)).step(time)
+
+  /**
+   * Step forward to `newTime`, evicting any wrapped buckets that
+   * fall outside of the window.
+   */
+  def step(newTime: Long): ExpHist =
+    if (newTime <= time) this
+    else {
+      val t = conf.expiration(newTime)
+      copy(
+        buckets = buckets.flatMap(_.expire(t)),
+        time = newTime)
+    }
+
+  /**
+   * "inc" algorithm:
+   *
+   *  * Step forward to the new timestamp
+   *  * create a new bucket with size 1 and timestamp
+   *  * Traverse the list of buckets in order of increasing sizes.
+   *
+   * If there are (k / 2 + 2) or more buckets of the same size, merge
+   * the oldest two of these buckets into a single bucket of double
+   * the size. Repeat until the bucket size is less than the limit.
+   */
+  def inc(timestamp: Long): ExpHist = {
+    val newBuckets =
+      buckets.lastOption match {
+        case None => Vector(BucketSeq.one(timestamp))
+        case Some(bucketSeq) => buckets.init :+ (bucketSeq + timestamp)
+      }
+    step(timestamp).copy(buckets = BucketSeq.normalize(conf.maxBuckets, newBuckets))
+  }
+
+  // Stupid implementation of `add` - just inc a bunch of times with
```

Comment:
we can kill this when we implement the next part of the paper - the `l-canonical` repr that'll let us calculate all of the bucket sizes at once and regenerate them.